File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/backend/AsmCode.cpp
#include "AsmCode.h"

using namespace Backend;

AsmCode::AsmCode(IR::Module *module)
    : module(module)
{
    for (auto globalVar : this->module->getGlobalVariableList())
    {
        globalVariableMap[globalVar] = new AsmGlobalVariable(globalVar);
    }

    for (auto irFunction : this->module->getFunctionList())
    {
        functionMap[irFunction] = new AsmFunction(this, irFunction);
    }

    for (auto &[key, value] : functionMap)
    {
        value->genFunction();
    }
}

AsmGlobalVariable *AsmCode::getGlobalVariable(IR::GlobalVariable *globalVar)
{
    if (globalVariableMap.find(globalVar) == globalVariableMap.end())
    {
        Error::Error(__PRETTY_FUNCTION__, "Global variable not found");
    }
    return globalVariableMap[globalVar];
}

AsmFunction *AsmCode::getFunction(IR::Function *irFunction)
{
    if (functionMap.find(irFunction) == functionMap.end())
    {
        Error::Error(__PRETTY_FUNCTION__, "Function not found");
    }
    return functionMap[irFunction];
}

AsmConstantFloat *AsmCode::getConstantFloat(float value)
{
    if (constFloatMap.find(value) == constFloatMap.end())
    {
        constFloatMap[value] = new AsmConstantFloat(value);
    }
    return constFloatMap[value];
}

AsmConstantLong *AsmCode::getConstantLong(long long value)
{
    if (constLongMap.find(value) == constLongMap.end())
    {
        constLongMap[value] = new AsmConstantLong(value);
    }
    return constLongMap[value];
}

std::string AsmCode::emit()
{
    std::string s;

    // Emit global variables
    for (auto [irGlobalVar, asmGlobalVar] : globalVariableMap)
        s += asmGlobalVar->emit();

    // Emit functions
    for (auto [irFunction, asmFunction] : functionMap)
        if (!irFunction->isBuiltinFunction())
            s += asmFunction->emit();

    // Emit constant values
    s += ".section .rodata\n";

    for (auto [key, value] : constFloatMap)
        s += value->emit();

    for (auto [key, value] : constLongMap)
        s += value->emit();

    return s;
}

File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/backend/asm/AsmCode.java
package cn.edu.bit.newnewcc.backend.asm;

import cn.edu.bit.newnewcc.ir.Module;
import cn.edu.bit.newnewcc.ir.value.BaseFunction;
import cn.edu.bit.newnewcc.ir.value.GlobalVariable;

import java.util.HashMap;
import java.util.Map;

public class AsmCode {
    private final Map<BaseFunction, AsmFunction> functionMap = new HashMap<>();
    private final Map<GlobalVariable, AsmGlobalVariable> globalVariableMap = new HashMap<>();
    private final Map<Float, AsmConstantFloat> constFloatMap = new HashMap<>();
    private final Map<Long, AsmConstantLong> constLongMap = new HashMap<>();

    public AsmGlobalVariable getGlobalVariable(GlobalVariable key) {
        return globalVariableMap.get(key);
    }

    public AsmConstantFloat getConstFloat(float value) {
        if (!constFloatMap.containsKey(value)) {
            constFloatMap.put(value, new AsmConstantFloat(value));
        }
        return constFloatMap.get(value);
    }
    public AsmConstantLong getConstLong(long value) {
        if (!constLongMap.containsKey(value)) {
            constLongMap.put(value, new AsmConstantLong(value));
        }
        return constLongMap.get(value);
    }
    public AsmCode(Module module) {
        for (var globalVariable : module.getGlobalVariables()) {
            AsmGlobalVariable variable = new AsmGlobalVariable(globalVariable);
            globalVariableMap.put(globalVariable, variable);
        }
        for (var function : module.getExternalFunctions()) {
            AsmFunction asmFunction = new AsmFunction(function, this);
            functionMap.put(function, asmFunction);
        }
        for (var function : module.getFunctions()) {
            AsmFunction asmFunction = new AsmFunction(function, this);
            functionMap.put(function, asmFunction);
        }
        for (var asmFunction : functionMap.values()) {
            asmFunction.emitCode();
        }
    }

    public AsmFunction getFunction(BaseFunction baseFunction) {
        return functionMap.get(baseFunction);
    }

    public String emit() {
        StringBuilder builder = new StringBuilder(".option nopic\n.attribute arch, \"rv64i2p1_m2p0_a2p1_f2p2_d2p2_c2p0_zicsr2p0_zifencei2p0_zba1p0_zbb1p0\"\n");
        builder.append(".attribute unaligned_access, 0\n");
        builder.append(".attribute stack_align, 16\n");
        for (var gvar : globalVariableMap.values()) {
            builder.append(gvar.emit());
        }
        for (var fvar : functionMap.values()) {
            if (!fvar.isExternal()) {
                builder.append(fvar.emit());
            }
        }
        builder.append(".section .rodata\n");
        for (var key : constFloatMap.keySet()) {
            var constFloat = constFloatMap.get(key);
            builder.append(constFloat.emit());
        }
        for (var key : constLongMap.keySet()) {
            var constLong = constLongMap.get(key);
            builder.append(constLong.emit());
        }
        return builder.toString();
    }
}
